### ISSUE TITLE
org.eclipse.xtext.xbase.tests.URIsInEcoreFilesTest fails in Eclipse

### ISSUE THREAD

**Issue by LorenzoBettini** [eclipse/xtext-extras#26]
@cdietrich I see that you modified org.eclipse.xtext.xbase.tests.AbstractPortableURIsTest in Commit 92101845df282d5ff1655b564c77ef27846b7381 but the test org.eclipse.xtext.xbase.tests.URIsInEcoreFilesTest fails in Eclipse with the following exception.

Can it be due to the fact that it creates platform URIs and this requires a running Eclipse? In that case it should be a UI test, right?

```
org.eclipse.xtext.resource.ClasspathUriResolutionException: org.eclipse.xtext.resource.FileNotFoundOnClasspathException: Couldn't find resource on classpath. URI was 'classpath:/model/Xbase.ecore'
    at org.eclipse.xtext.resource.ClassloaderClasspathUriResolver.resolve(ClassloaderClasspathUriResolver.java:49)
    at org.eclipse.xtext.resource.XtextResourceSet.resolveClasspathURI(XtextResourceSet.java:221)
    at org.eclipse.xtext.resource.XtextResourceSet.access$0(XtextResourceSet.java:220)
    at org.eclipse.xtext.resource.XtextResourceSet$1.normalize(XtextResourceSet.java:284)
    at org.eclipse.emf.ecore.resource.impl.ExtensibleURIConverterImpl.normalize(ExtensibleURIConverterImpl.java:449)
    at org.eclipse.xtext.resource.XtextResourceSet$1.normalize(XtextResourceSet.java:290)
    at org.eclipse.xtext.resource.XtextResourceSet.getResource(XtextResourceSet.java:239)
    at org.eclipse.xtext.xbase.tests.AbstractPortableURIsTest.doTestResource(AbstractPortableURIsTest.java:89)
    at org.eclipse.xtext.xbase.tests.URIsInEcoreFilesTest.testXbase(URIsInEcoreFilesTest.java:18)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
    at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
Caused by: org.eclipse.xtext.resource.FileNotFoundOnClasspathException: Couldn't find resource on classpath. URI was 'classpath:/model/Xbase.ecore'
    at org.eclipse.xtext.resource.ClassloaderClasspathUriResolver.findResourceOnClasspath(ClassloaderClasspathUriResolver.java:61)
    at org.eclipse.xtext.resource.ClassloaderClasspathUriResolver.resolve(ClassloaderClasspathUriResolver.java:46)
    ... 33 more
```

**Comment by cdietrich** [comment]
sry i have no idea on that. this has to be investigated. unfortunately the classpath regarding model/ is different in gradle and eclipse. maybe a issue with buildship.

in buildship model/* is added to the jar and this makes it part of the classpath.
i guess someone who knows gradle should have a look

**Comment by LorenzoBettini** [comment]
Exactly the same problem here: org.eclipse.xtext.xtext.URIsInEcoreFilesTest

**Comment by cdietrich** [comment]
@oehme do you have an idea how that could be solved?

**Comment by LorenzoBettini** [comment]
I don't know gradle, but if it behaves like tycho-surefire-plugin then it executes the tests in an OSGI environment (without the UI) and this would make platform URIs work

**Comment by cdietrich** [comment]
no, these uris actually work. the problem is that the classpath uri does not work cause model/* is not on classpath in eclipse. same for org/..../Xtext.ecore/genmodel.

these files are jared like

```
jar {
    from('.') {
        include 'org/**', 'modeling32.png'
    }
}

```

or

```
jar {
	from ('.') {
		include 'about*.*', 'plugin.xml', 'schema/**', 'model/**'
	}
	metaInf {
		from 'META-INF'
	}
}

```

**Comment by LorenzoBettini** [comment]
but these tests used to work before the repo splitting, didn't they? I guess because at that time they were Eclipse plug-in projects (PDE nature). Could that be?

Or probably setting 'model' as a source folder would do the trick, because then their contents would be copied to the bin folder?

**Comment by cdietrich** [comment]
I did not look at the old state. I don't know why this is done this way / why the files are not in a src folder

**Comment by cdietrich** [comment]
For setting model as source folder the places that look for model/xxxx.ecore/genmodel would have to be changed

**Comment by oehme** [comment]
This is indeed caused by switching off the plugin nature. The `model` folder was added to the output of the project through a declaration in build.properties.

You cannot make `model` a source folder because of how platform resource URIs work: They require the resource to have the same relative path inside the workspace project and the final jar file.

The only solution I can think of is to add the project root folder as a source folder with a single include for `model`. For now you'd have to check in that customization in your .classpath file. With Gradle 3.0/Buildship 1.0.18 (both out soon), you'll be able to automate this customization.

**Comment by LorenzoBettini** [comment]
and what about simply adding the plugin nature back? It will be used only when opening the project in Eclipse and it should not disturb anything else, I guess.

**Comment by oehme** [comment]
I see no problem with that. I was only offering the alternative if for some reason you really don't want the plugin nature :)

**Comment by oehme** [comment]
Oh another obvious solution I missed: Use classpath URIs and put the model in `src/main/resources`.

**Comment by cdietrich** [comment]
ill try to fix this using a custom urlclassloader

**Comment by cdietrich** [comment]
Fixed in 2.13
